scylla/src/transport/caching_session.rs

```diff
@@ -0,0 +1,272 @@
+                self.session.execute(&new_prepared_statement, values).await
+            }
+        }
+    }
+
+
+    /// Does the same thing as [`Session::execute_iter`] but uses the prepared statement cache
+    pub async fn execute_iter(
+        &self,
+        query: impl Into<Query>,
+        values: impl ValueList,
+    ) -> Result<RowIterator, QueryError> {
+        let query = query.into();
+        let prepared = self.add_prepared_statement(&query).await?;
+        let values = values.serialized()?;
+        let result = self.session.execute_iter(prepared, values.clone()).await;
+
+        match self.post_execute_prepared_statement(&query, result).await? {
+            Either::Left(result) => Ok(result),
+            Either::Right(new_prepared_statement) => {
+                self.session.execute_iter(new_prepared_statement, values).await
+            }
+        }
+    }
+
+    /// Does the same thing as [`Session::execute_paged`] but uses the prepared statement cache
+    pub async fn execute_paged(
+        &self,
+        query: impl Into<Query>,
+        values: impl ValueList,
+        paging_state: Option<Bytes>,
+    ) -> Result<QueryResult, QueryError> {
+        let query = query.into();
+        let prepared = self.add_prepared_statement(&query).await?;
+        let values = values.serialized()?;
+        let result = self
+            .session
+            .execute_paged(&prepared, values.clone(), paging_state.clone())
+            .await;
+
+        match self.post_execute_prepared_statement(&query, result).await? {
+            Either::Left(result) => Ok(result),
+            Either::Right(new_prepared_statement) => {
+                self
+                    .session
+                    .execute_paged(&new_prepared_statement, values, paging_state)
+                    .await
+            }
+        }
+    }
+
+    /// Adds a prepared statement to the cache
+    pub async fn add_prepared_statement(
+        &self,
+        query: impl Into<&Query>,
+    ) -> Result<PreparedStatement, QueryError> {
+        let query = query.into();
+
+        if let Some(prepared) = self.cache.get(&query.contents) {
+            // Clone, because else the value is mutably borrowed and the execute method gives a compile error
+            Ok(prepared.clone())
+        } else {
+            let prepared = self.session.prepare(query.clone()).await?;
+
+            if self.max_capacity
+                == self.cache.len()
+            {
+                // Cache is full, remove the first entry
+                // Don't delete while holding the key, this could deadlock
+                // Instead, store the raw string in a variable and remove it later on when there
+                // are no more references to the cache
+                let mut query = "".to_string();
+
+                // There is no easy way to just get a random/first element
+                #[allow(clippy::never_loop)]
+                for first in self.cache.iter() {
```

Comment:
[`.next()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#tymethod.next) could be used instead of `for` loop here.

Reply:
Thanks for the suggestion, didn't knew that one! I made the code changes